File: Python3/318. Maximum Product of Word Lengths.py
class Solution:
    def maxProduct(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        wordDict = {} # 位运算, elegant的想法
        for word in words:
            mask = 0
            for char in word:
                mask |= 1 << (ord(char) - ord('a'))
            wordDict[mask] = max(wordDict.get(mask, 0), len(word))
        ans = 0
        for mask1, mask2 in itertools.combinations(wordDict.keys(), 2):
            if not (mask1 & mask2):
                ans = max(ans, wordDict[mask1] * wordDict[mask2])
        return ans

File: Golang/318. Maximum Product of Word Lengths.go
func maxProduct(words []string) int {
    lenDict := make(map[uint]int, len(words))
    for _, word := range words {
        var mask uint
        for _, char := range word {
            mask |= uint(1) << uint(char - 'a')
        }
        if len(word) > lenDict[mask] {
            lenDict[mask] = len(word)
        }
    }
    ans := 0
    for item := range lenDict {
        for item2 := range lenDict {
            if item & item2 == 0 {
                if lenDict[item] * lenDict[item2] > ans {
                    ans = lenDict[item] * lenDict[item2]
                }
            }
        }
    }
    return ans
}
